app/helpers/application_helper.rb

```diff
@@ -15,4 +16,40 @@ def like_or_dislike_btn(post)
       link_to('Like!', post_likes_path(post_id: post.id), method: :post)
     end
   end
+
+  def pending_friendships
+    start = false
+    Friendship.all.each do |a|
+      if a.confirmed == false && current_user.id == a.user_id
+        start = true
+      end
+    end
+  end
+
+  def incoming_friendships
+    current_user.friendships.each do |a|
+      if a.confirmed == false && current_user.id = a.user_id
+        true
+      end
+    end
+  end
+
+  def already_sent_request_or_friends(current_user)
+    a = Friendship.where('confirmed =?', false).pluck(:friend_id, :user_id).flatten
+    b = Friendship.where('confirmed =?', true).pluck(:friend_id, :user_id).flatten
+    if a.include?(current_user.id) || b.include?(current_user.id)
+      true
+    else
+      false
+    end
+  end
+
+  def already_friends(current_user)
+    b = Friendship.where('confirmed =?', true).pluck(:friend_id, :user_id).flatten
+    if b.include?(current_user.id)
+      true
+    else
+      false
+    end
+  end
```

Comment:
- [ ] Once a friendship is confirmed, an additional row with reversed user-friend pair should be created.
```
# Friendship model

def confirm_friend
  self.update_attributes(confirmed: true)
  Friendship.create!(friend_id: self.user_id,
                  user_id: self.friend_id,
                  confirmed: true)
end

```

- [ ] With the additional row, we can then easily find friends only by user_id in the friendships table.
```
# User model

has_many :confirmed_friendships, -> { where confirmed: true }, class_name: "Friendship"
has_many :friends, through: :confirmed_friendships

```

- [ ] A user's pending_friends and friend_requests can be defined as well as association with SQL query instead of processing objects in application memory (as the database was created to store and query data it is always more efficient than ruby).
```
# User model

# Users who needs to confirm friendship
has_many :pending_friendships, -> { where confirmed: false }, class_name: "Friendship", foreign_key: "user_id"
has_many :pending_friends, through: :pending_friendships, source: :friend

# Users who requested to be friends (needed for notifications)
has_many :inverted_friendships, -> { where confirmed: false }, class_name: "Friendship", foreign_key: "friend_id"
has_many :friend_requests, through: :inverted_friendships


```

- [ ] Currently, users who are not friends can all view all users' posts even when they are not friends. The expected behavior is user A can only view their own posts as well as their friends' posts (say user B)
```
# User model

def friends_and_own_posts
  Post.where(user: (self.friends.to_a << self))
  # This will produce SQL query with IN. Something like: select * from posts where user_id IN (1,45,874,43);
end
```